blockstack-vm/src/types.rs

```diff
@@ -0,0 +1,213 @@
+                return Err(Error::InvalidArguments("Cannot use named argument twice in tuple construction.".to_string()))
+            }
+        }
+        Ok(TupleTypeSignature { type_map: type_map })
+    }
+
+    pub fn check_valid(&self, name: &str, value: &Value) -> bool {
+        if let Some(expected_type) = self.type_map.get(name) {
+            *expected_type == TypeSignature::type_of(value)
+        } else {
+            false
+        }
+    }
+}
+
+impl TupleData {
+    pub fn from_data(data: &[(&str, Value)]) -> Result<TupleData, Error> {
+        let mut type_map = BTreeMap::new();
+        let mut data_map = BTreeMap::new();
+        for (name, value) in data {
+            let type_info = TypeSignature::type_of(value);
+            if type_info.atomic_type == AtomTypeIdentifier::VoidType {
+                return Err(Error::InvalidArguments("Cannot use VoidTypes in tuples.".to_string()))
+            }
+            if let Some(_v) = type_map.insert(name.to_string(), type_info) {
+                return Err(Error::InvalidArguments("Cannot use named argument twice in tuple construction.".to_string()))
+            }
+            data_map.insert(name.to_string(), (*value).clone());
+        }
+        Ok(TupleData { type_signature: TupleTypeSignature { type_map: type_map },
+                       data_map: data_map })
+
+    }
+
+    pub fn get(&self, name: &str) -> InterpreterResult {
+        if let Some(value) = self.data_map.get(name) {
+            Ok(value.clone())
+        } else {
+            Err(Error::InvalidArguments(format!("No such field {:?} in tuple", name)))
+        }
+        
+    }
+}
+
+impl TypeSignature {
+    pub fn new(atomic_type: AtomTypeIdentifier, dimension: u8) -> TypeSignature {
+        TypeSignature { atomic_type: atomic_type,
+                        dimension: dimension }
+    }
+
+    pub fn type_of(x: &Value) -> TypeSignature {
+        match x {
+            Value::Void => TypeSignature::new(AtomTypeIdentifier::VoidType, 0),
+            Value::Int(_v) => TypeSignature::new(AtomTypeIdentifier::IntType, 0),
+            Value::Bool(_v) => TypeSignature::new(AtomTypeIdentifier::BoolType, 0),
+            Value::Buffer(_v) => TypeSignature::new(AtomTypeIdentifier::BufferType, 0),
+            Value::List(_v, type_signature) => type_signature.clone(),
+            Value::Tuple(v) => TypeSignature::new(AtomTypeIdentifier::TupleType(
+                v.type_signature.clone()), 0)
+        }
+    }
+
+    pub fn get_list_type_for(x: &Value) -> Result<TypeSignature, Error> {
+        match x {
+            Value::Void => Err(Error::InvalidArguments("Cannot construct list of void types".to_string())),
+            Value::Tuple(_a) => Err(Error::InvalidArguments("Cannot construct list of tuple types".to_string())),
+            _ => {
+                let mut base_type = TypeSignature::type_of(x);
+                base_type.dimension += 1;
```

Comment:
Shouldn't this be a `checked_add`?  What happens if the dimension overflows?

Reply:
Yep, good catch, thanks